rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -190,3 +190,121 @@ pub fn process_action_schema(
         actions.insert(#name_str.to_string(), #psibase_mod::fracpack::FunctionType{ params: builder.insert::<#action_mod::#name>(), result: #ret });
     }
 }
+
+pub struct PreAction {
+    pub exists: bool,
+    pub fn_name: Option<Ident>,
```

Comment:
Is there ever a case where `exists != fn_name.is_some()`?

Reply:
nope.
I turned exists: bool into a helper function has_pre_action()